blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,361 @@
+certificates
+- MOK (Machine Owner Key) Secure Boot, which extends UEFI SB by
+introducing user-added Machine Owner Keys
+
+### Custom certificates
+
+UEFI Secure Boot allows the user to enroll custom certificates which will be
+used to verify images. The certificates can be loaded automatically thanks to
+the
+[Automatic Certificate Provisioning](https://github.com/Wind-River/meta-secure-core/tree/master/meta-efi-secure-boot#automatic-certificate-provision)
+procedure provided by `meta-secure-core`. The procedure uses `LockDown.efi`,
+which can only be executed when UEFI Secure Boot is disabled. `LockDown.efi`
+provisions the new certificates, which are built into it during build.
+
+UEFI Secure Boot uses the DB certificate to verify GRUB. Aside from that,
+`grubenv` and `grub.cfg` are signed during build using a boot key, which can
+be defined in the layer. GRUB then tries to load the kernel, which is also
+checked against DB. If any of the checks fail, the respective file cannot be
+booted. This mechanism increases the system's security by making sure that the
+firmware does not boot untrusted files.
+
+### Implementation
+
+To integrate the mechanisms available in `meta-secure-core` you need to
+integrate that layer into your build. We use
+[kas-container](https://github.com/siemens/kas/blob/master/kas-container) to set
+up bitbake projects.
+
+In your kas configuration file add the following to the `repos` section:
+
+```yaml
+meta-secure-core:
+url: https://github.com/Wind-River/meta-secure-core.git
+refspec: 8dc9f1b4a735eccee65a8896760e473e110d147e
+layers:
+    meta-efi-secure-boot:
+    meta:
+    meta-signing-key:
+```
+
+Define `meta-secure-core` variables in your layer's `local.conf`:
+
+```conf
+# UEFI Secure Boot variables
+# Use only UEFI Secure Boot without MOK Secure Boot
+UEFI_SELOADER = "0"
+GRUB_SIGN_VERIFY = "1"
+UEFI_SB = "1"
+MOK_SB = "0"
+
+# We want grub-efi from meta-efi-secure-boot to install bootfiles under
+# /boot/EFI/BOOT
+EFI_BOOT_PATH = "/boot/EFI/BOOT"
+
+DISTRO_FEATURES_NATIVE:append = " efi-secure-boot"
+DISTRO_FEATURES:append = " efi-secure-boot modsign"
+MACHINE_FEATURES_NATIVE:append = " efi"
+MACHINE_FEATURES:append = " efi"
+
+DEBUG_FLAGS:forcevariable = ""
+IMAGE_INSTALL:append = " kernel-image-bzimage"
+```
+
+Also, make sure to install the `efi-secure-boot` packagegroup into your image:
+
+```bitbake
+IMAGE_INSTALL:append = " \
+    packagegroup-efi-secure-boot \
+"
+```
+
+We were using a custom version of the `bootimg-efi` wic plugin to set up a
+UEFI-compliant image. That required us to install UEFI Secure Boot-related files
+in a specific directory:
+
+```bitbake
+# grub-efi_%.bbappend
+
+do_deploy:append:class-target () {
+    # deploy all UEFI SB related bootfiles to later use them with bootimg-efi
+    # wic plugin
+    install -d ${DEPLOY_DIR_IMAGE}/bootfiles
+    cp -r ${D}/boot/* ${DEPLOY_DIR_IMAGE}/bootfiles/
+}
+```
+
+We also needed to define the
```

Comment:
@PLangowski I would rather say that used wic plugin needs IMAGE_BOOT_FILES defined so for this reason below entry was added

Reply:
ad87522d2cb6e682dc5e76ebb3c26599f8ec764c